### DISCUSSION TITLE
Does the EntityRepository differ from the EntityManager about Database Connection?

### DISCUSSION THREAD

**Discussion by jays1204** [mikro-orm/mikro-orm#3030] [Q&A]
I read this page https://mikro-orm.io/docs/repositories. And I understood EntityRepository is no different from EntityManager.

But below code show EntityRepository has some different.

example code, Database connection configuration has min:  1, max: 1. and database is Mysql

MyService.ts
```ts
@Injectable()
export class MyService {
  constructor(
    @InjectEntityManager()
    private readonly entityManager: EntityManager,
    private readonly mydao: MyDao
  ) {}


  async myMethod() {
    await this.entityManager.transactional(async em => {
       await this.myDao.findMyPost('THIS_WILL_NOT_BE_SAVE');
    });
  }
.... 
}

```

MyDao.ts
```ts
@Injectable()
export class MyDao {
  constructor(
    @InjectEntityManager()
    private readonly entityManager: EntityManager
  ) {}

   async findMyPost(text) {
    const post = new Post(); // Post is Entity
    post.content = text;
    await this.entityManager.findOne(Post, {text}); // error
    await this.entityManager.getRepository(Post).find({text}); // works
  }

```
When I call myMethod,

`await this.entityManager.findOne(Post, {text});` this statement occur error message like this. `DriverException: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call`

But `await this.entityManager.getRepository(Post).find({text})` method works well.

`this.entityManager.getRepository` keep database connection, but `this.entityManager` has another database connection.

Is it right? Or did it happen by accident?

**Comment by B4nan** [comment]
No, they are not different, repository just wraps EM, it literally uses EM to do things. Those two are equal:

```ts
await this.entityManager.findOne(Post, {text});
await this.entityManager.getRepository(Post).find({text});
```

You probably have a problem with contexts, and both of those calls should end up using `em.getContext()` which should use the contextual EM. Are you on v5? Is this happening inside an explicit transaction? Do you use `RequestContext` helper? It's nest app, right? 

Might be a problem with `@InjectEntityManager` decorator - note that you don't need that unless you have multiple ORM instances in a single project (and there you need to use its parameter to select the EM by name).

**Reply by jays1204** [reply]
Yes My app is nest.js and Mikro-ORM is v5. And I do not use RequestContext.  and this happens inside an explicit transaction.

As you mentioned it, I will try test after remove InjectEntityManager.
I expect both works well.
```ts
await this.entityManager.findOne(Post, {text});
await this.entityManager.getRepository(Post).find({text});
```

`this.entityManager.getRepository` and `this.entityManager` keep database connection. is it right?

**Reply by B4nan** [reply]
Sort of, the database connection is kept in the driver instance, and that is shared - you only have one instance. We have then one instance of a connection class for each replica (so one by default if you dont use replication), and it holds a single knex instance - knex is managing connection pooling for us (via another dependency called tarn). The problem with acquiring the connection usually means you work with a wrong EM isntance, as EM is what holds the transaction context, and if the transaction is already closed, you cannot use it for other queries.

If you see this inside explicit transaction, the `em.getContext()` call I mentioned earlier should always prefer a transactional EM fork that is available inside this transaction handler. Note that you can compare EM instances via their `id` property (to check on repository instance you will need to cast to `any` first as the `em` getter is protected, so `(repo as any).em.id`):

```ts
console.log(this.entityManager.id);
console.log((this.entityManager.getRepository(Post) as any).em.id);
```

**Reply by jays1204** [reply]
I tested by removed InjectEntityManager and multiple database configuration.
But still error works. (every em.id is same)

below repository show error.
https://github.com/jays1204/mikro-test ( this repository forks https://github.com/mikro-orm/nestjs-realworld-example-app. fix some database connection config, and POST /api/articles API)

If you call POST /api/articles, then you can see timeout error. 
Because this line. https://github.com/jays1204/mikro-test/blob/master/src/article/article.service.ts#L95

comment https://github.com/jays1204/mikro-test/blob/master/src/article/article.service.ts#L95 line and uncomment https://github.com/jays1204/mikro-test/blob/master/src/article/article.service.ts#L97, Then works well.

I expected that entityManager(injected on constructor, like https://github.com/jays1204/mikro-test/blob/master/src/article/article.service.ts#L22) get proper connection in  explicit transaction, but it is not.
Otherwise Repository get proper connection well.

Did I use it wrong?

I want to call method of another class in explicit transaction and keep connection in single explicit transaction callback.

**Reply by jays1204** [reply]
@B4nan  is this a bug?

**Reply by B4nan** [reply]
if its a bug, then its a bug with the nestjs adapter, not with the ORM itself (=> wrong repository). i didnt have time to look into this more yet.

> But still error works. (every em.id is same)

I have no idea what you mean by this. if `em.id` is same, that is the correct thing (unless its `1` which is the global EM, but you would get a validation warning in v5 for that). they should be the same, as they shuold all use the same context. that is correct behaviour, not error. and saying "error works", what does that even mean? :P

**Reply by jays1204** [reply]
I am sorry for my bad english. 
I mean still error occur. 
em.id same on https://github.com/jays1204/mikro-test/blob/master/src/article/article.service.ts#L95 and  https://github.com/jays1204/mikro-test/blob/master/src/article/article.controller.ts#L51.

**Reply by B4nan** [reply]
Yes, and that is correct, they should be the same. Everything inside the `transactional` handler should have the same `em.id`.

**Reply by B4nan** [reply]
Try to prepare _minimal_ reproduction, I don't have time to debug such a complex app.

**Reply by jays1204** [reply]
Ok. I will inform to you after prepare minimal reproduction.
For your information, [this example repository](https://github.com/jays1204/mikro-test) is forked from [your repository](https://github.com/mikro-orm/nestjs-realworld-example-app)

**Reply by B4nan** [reply]
Yes, and that is far away from what you can call a _minimal_ reproduction. Its actually the very opposite, its a _realworld_ example.
